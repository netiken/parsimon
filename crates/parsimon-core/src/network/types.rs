use std::cmp::Ordering;

use crate::edist::EDistBuckets;
use crate::units::{Bytes, Gbps, Nanosecs};

pub const PKTSIZE_MAX: Bytes = Bytes::new(1000);

#[derive(Debug, Clone, PartialEq, Eq, serde::Serialize, serde::Deserialize)]
pub struct Node {
    pub id: NodeId,
    pub kind: NodeKind,
}

impl Node {
    pub fn new_host(id: NodeId) -> Self {
        Self {
            id,
            kind: NodeKind::Host,
        }
    }

    pub fn new_switch(id: NodeId) -> Self {
        Self {
            id,
            kind: NodeKind::Switch,
        }
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, serde::Serialize, serde::Deserialize)]
pub enum NodeKind {
    Host,
    Switch,
}

identifier!(NodeId, usize);

/// A bidirectional channel.
#[derive(Debug, Clone, Copy, derive_new::new, serde::Serialize, serde::Deserialize)]
pub struct Link {
    pub a: NodeId,
    pub b: NodeId,
    pub bandwidth: Gbps,
    pub delay: Nanosecs,
}

impl Link {
    pub fn connects(&self, x: NodeId, y: NodeId) -> bool {
        self.a == x && self.b == y || self.a == y && self.b == x
    }
}

pub trait Channel {
    fn src(&self) -> NodeId;

    fn dst(&self) -> NodeId;

    fn bandwidth(&self) -> Gbps;

    fn delay(&self) -> Nanosecs;
}

// All channels just copy these fields
macro_rules! channel_impl {
    ($name: ty) => {
        impl Channel for $name {
            fn src(&self) -> NodeId {
                self.src
            }

            fn dst(&self) -> NodeId {
                self.dst
            }

            fn bandwidth(&self) -> Gbps {
                self.bandwidth
            }

            fn delay(&self) -> Nanosecs {
                self.delay
            }
        }
    };
}

#[derive(Debug, Clone, PartialEq, Eq, derive_new::new, serde::Serialize)]
pub(crate) struct BasicChannel {
    pub(crate) src: NodeId,
    pub(crate) dst: NodeId,
    pub(crate) bandwidth: Gbps,
    pub(crate) delay: Nanosecs,
}

channel_impl!(BasicChannel);

#[derive(Debug, Clone, PartialEq, Eq, serde::Serialize)]
pub struct FlowChannel {
    pub(crate) src: NodeId,
    pub(crate) dst: NodeId,
    pub(crate) bandwidth: Gbps,
    pub(crate) delay: Nanosecs,
    pub(crate) flows: Vec<FlowId>,
}

channel_impl!(FlowChannel);

impl FlowChannel {
    pub(crate) fn new_from(chan: &BasicChannel) -> Self {
        Self {
            src: chan.src,
            dst: chan.dst,
            bandwidth: chan.bandwidth,
            delay: chan.delay,
            flows: Vec::new(),
        }
    }

    /// Get an iterator over the traced channel's flow IDs
    pub fn flow_ids(&self) -> impl Iterator<Item = FlowId> + '_ {
        self.flows.iter().copied()
    }

    delegate::delegate! {
        to self.flows {
            #[call(len)]
            pub fn nr_flows(&self) -> usize;

            #[call(push)]
            pub(crate) fn push_flow(&mut self, flow: FlowId);
        }
    }
}

#[derive(Debug, Clone)]
#[allow(unused)]
pub struct EDistChannel {
    pub(crate) src: NodeId,
    pub(crate) dst: NodeId,
    pub(crate) bandwidth: Gbps,
    pub(crate) delay: Nanosecs,
    pub(crate) dists: EDistBuckets,
}

impl EDistChannel {
    pub(crate) fn new_from(chan: &FlowChannel) -> Self {
        Self {
            src: chan.src,
            dst: chan.dst,
            bandwidth: chan.bandwidth,
            delay: chan.delay,
            dists: EDistBuckets::new_empty(),
        }
    }
}

channel_impl!(EDistChannel);

#[derive(Debug)]
pub struct Path<'a, C> {
    inner: Vec<&'a C>,
}

impl<'a, C: Channel> Path<'a, C> {
    pub fn new(channels: Vec<&'a C>) -> Self {
        Self { inner: channels }
    }

    pub fn delay(&self) -> Nanosecs {
        self.inner.iter().map(|&c| c.delay()).sum()
    }

    pub fn bandwidths(&self) -> impl Iterator<Item = Gbps> + '_ {
        self.inner.iter().map(|&c| c.bandwidth())
    }
}

identifier!(FlowId, usize);

#[derive(Debug, Default, Clone, Copy, Hash, serde::Serialize, serde::Deserialize)]
pub struct Flow {
    pub id: FlowId,
    pub src: NodeId,
    pub dst: NodeId,
    pub size: Bytes,
    pub start: Nanosecs,
}

#[derive(Debug, Clone, Copy)]
pub struct FctRecord {
    // From flow
    pub id: FlowId,
    pub size: Bytes,
    pub start: Nanosecs,
    // From simulation
    pub fct: Nanosecs,
    pub ideal: Nanosecs,
}

impl FctRecord {
    pub fn delay(&self) -> Nanosecs {
        // Work around kind-of-wrong ns-3 computation
        match self.fct.cmp(&self.ideal) {
            Ordering::Less | Ordering::Equal => Nanosecs::ZERO,
            Ordering::Greater => self.fct - self.ideal,
        }
    }

    pub fn pktnorm_delay(&self) -> f64 {
        let nr_pkts = (self.size.into_f64() / PKTSIZE_MAX.into_f64()).ceil();
        self.delay().into_f64() / nr_pkts
    }
}
