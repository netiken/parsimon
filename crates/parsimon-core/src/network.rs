use petgraph::graph::DiGraph;

#[derive(Debug)]
pub struct Network {
    topology: DiGraph<Node, Channel>,
    routes: (), // copy this from cloudburst
}

impl Network {}

#[derive(Debug)]
struct Node;

#[derive(Debug)]
struct Channel;

#[derive(Debug)]
pub struct DelayNet;
