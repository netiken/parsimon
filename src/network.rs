// A topology that we know is valid.
// Includes topology and routes
// - This is probably enough to replicate ECMP decisions
#[derive(Debug)]
pub struct Network {
    topology: (), // maybe this should just be a graph
    routes: (),   // copy this from cloudburst
}

impl Network {}
