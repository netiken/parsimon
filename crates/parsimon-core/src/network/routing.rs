use std::collections::{HashMap, VecDeque};

use petgraph::{
    graph::NodeIndex,
    visit::{VisitMap, Visitable},
};

use crate::network::{
    topology::Topology,
    types::{Channel, NodeId, NodeKind},
};

pub(super) type HopMatrix = HashMap<NodeId, HopMap>;
pub(super) type HopMap = HashMap<NodeId, Vec<NodeId>>;

#[derive(Debug, Clone, serde::Serialize)]
pub(super) struct Routes {
    inner: HopMatrix,
}

impl Routes {
    /// Builds a routing table from a topology using BFS.
    pub(super) fn new(topology: &Topology<Channel>) -> Self {
        let g = &topology.graph;
        let mut hops = HopMatrix::new();
        for start in g.node_indices() {
            let mut discovered = g.visit_map();
            discovered.visit(start);

            let mut queue = VecDeque::new();
            queue.push_back(start);

            let mut distances: HashMap<NodeIndex, usize> = [(start, 0)].into_iter().collect();

            while let Some(n) = queue.pop_front() {
                let cur_distance = *distances.get(&n).unwrap();
                for succ in g.neighbors(n) {
                    if discovered.visit(succ) {
                        distances.insert(succ, cur_distance + 1);
                        if matches!(g[succ].kind, NodeKind::Switch) {
                            queue.push_back(succ);
                        }
                    }
                    // In this function, we do not assume `NodeId`s and `NodeIndex`s are exactly
                    // the same, but it may be enforced elsewhere
                    if *distances.get(&succ).unwrap() == cur_distance + 1 {
                        hops.entry(g[succ].id)
                            .or_default()
                            .entry(g[start].id)
                            .or_default()
                            .push(g[n].id);
                    }
                }
            }
        }
        Self { inner: hops }
    }

    pub(super) fn for_node(&self, node: NodeId) -> Option<&HopMap> {
        self.inner.get(&node)
    }

    pub(super) fn next_hops_unchecked(&self, from: NodeId, to: NodeId) -> &[NodeId] {
        self.for_node(from)
            .expect("missing node in routes")
            .get(&to)
            .expect("missing route to node")
    }
}

#[cfg(test)]
mod tests {
    use std::collections::BTreeMap;

    use super::*;
    use crate::testing;
    use anyhow::Context;

    type SortedHopMatrix = BTreeMap<NodeId, SortedHopMap>;
    type SortedHopMap = BTreeMap<NodeId, Vec<NodeId>>;

    /// Generate a stable sorting of the hop matrix for snapshot tests
    fn sorted_hop_matrix(matrix: &HopMatrix) -> SortedHopMatrix {
        matrix
            .iter()
            .map(|(&id, m)| {
                let m = m
                    .iter()
                    .map(|(&id, hops)| {
                        let mut hops = hops.clone();
                        hops.sort();
                        (id, hops)
                    })
                    .collect::<BTreeMap<_, _>>();
                (id, m)
            })
            .collect::<BTreeMap<_, _>>()
    }

    #[test]
    fn route_three_node_succeeds() -> anyhow::Result<()> {
        let topo = testing::three_node_topology().context("failed to create topology")?;
        let routes = Routes::new(&topo);
        let hops = sorted_hop_matrix(&routes.inner);
        insta::assert_yaml_snapshot!(hops);
        Ok(())
    }

    #[test]
    fn route_eight_node_succeeds() -> anyhow::Result<()> {
        let topo = testing::eight_node_topology().context("failed to create topology")?;
        let routes = Routes::new(&topo);
        let hops = sorted_hop_matrix(&routes.inner);
        insta::assert_yaml_snapshot!(hops);
        Ok(())
    }
}
