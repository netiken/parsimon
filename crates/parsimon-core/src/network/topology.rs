use std::collections::{HashMap, HashSet};

use petgraph::graph::{DiGraph, EdgeIndex, NodeIndex};

use crate::network::types::{Channel, Link, Node, NodeId, NodeKind, TracedChannel};

use super::types::EDistChannel;

#[derive(Debug)]
pub(crate) struct Topology<C> {
    pub(crate) graph: DiGraph<Node, C>,
    pub(crate) id2idx: HashMap<NodeId, NodeIndex>,
    pub(crate) links: Vec<Link>,
}

impl<C> Topology<C> {
    delegate::delegate! {
        to self.id2idx {
            #[call(get)]
            pub(crate) fn idx_of(&self, id: &NodeId) -> Option<&NodeIndex>;
        }

        to self.graph {
            #[allow(unused)]
            #[call(edge_count)]
            pub(crate) fn nr_edges(&self) -> usize;

            pub(crate) fn find_edge(&self, a: NodeIndex, b: NodeIndex) -> Option<EdgeIndex>;
        }
    }
}

impl Topology<Channel> {
    /// Creates a network topology from a list of nodes and links. This function returns an error if
    /// the given specification fails to produce a valid topology. The checks are not exhaustive.
    ///
    /// Correctness properties:
    ///
    /// - Every node must have a unique ID.
    /// - Every link must have distinct endpoints in `nodes`.
    /// - Every node must be referenced by some link.
    /// - For any two nodes, there must be at most one link between them.
    /// - Every host node should only have one link.
    pub(crate) fn new(nodes: &[Node], links: &[Link]) -> Result<Self, TopologyError> {
        let mut g = DiGraph::new();
        let mut id2idx = HashMap::new();
        for n @ Node { id, .. } in nodes.iter().cloned() {
            let idx = g.add_node(n);
            if id2idx.insert(id, idx).is_some() {
                // CORRECTNESS: Every node must have a unique ID.
                return Err(TopologyError::DuplicateNodeId(id));
            }
        }
        let idx_of = |id| *id2idx.get(&id).unwrap();
        let mut referenced_nodes = HashSet::new();
        for Link {
            a,
            b,
            bandwidth,
            delay,
        } in links.iter().cloned()
        {
            // CORRECTNESS: Every link must have distinct endpoints in `nodes`.
            if a == b {
                return Err(TopologyError::NodeAdjacentSelf(a));
            }
            if !id2idx.contains_key(&a) {
                return Err(TopologyError::UndeclaredNode(a));
            }
            if !id2idx.contains_key(&b) {
                return Err(TopologyError::UndeclaredNode(b));
            }
            referenced_nodes.insert(a);
            referenced_nodes.insert(b);
            // Channels are unidirectional
            g.add_edge(idx_of(a), idx_of(b), Channel::new(a, b, bandwidth, delay));
            g.add_edge(idx_of(b), idx_of(a), Channel::new(b, a, bandwidth, delay));
        }
        // CORRECTNESS: Every node must be referenced by some link.
        for &id in id2idx.keys() {
            if !referenced_nodes.contains(&id) {
                return Err(TopologyError::IsolatedNode(id));
            }
        }
        for eidx in g.edge_indices() {
            // CORRECTNESS: For any two nodes, there must be at most one link between them.
            let (a, b) = g.edge_endpoints(eidx).unwrap();
            if g.edges_connecting(a, b).count() > 1 {
                return Err(TopologyError::DuplicateLink {
                    n1: g[a].id,
                    n2: g[b].id,
                });
            }
            // CORRECTNESS: Every host node should only have one link.
            let Node { id, kind, .. } = g[a];
            if matches!(kind, NodeKind::Host) {
                let nr_outgoing = g.edges(a).count();
                if nr_outgoing > 1 {
                    return Err(TopologyError::TooManyHostLinks { id, n: nr_outgoing });
                }
            }
        }
        Ok(Self {
            graph: g,
            id2idx,
            links: Vec::from(links),
        })
    }
}

impl Topology<TracedChannel> {
    pub(crate) fn new_traced(topology: &Topology<Channel>) -> Self {
        // CORRECTNESS: For nodes and edges, `petgraph` guarantees that the
        // iteration order matches the order of indices.
        let mut g = DiGraph::new();
        for node in topology.graph.node_weights() {
            g.add_node(node.clone());
        }
        for eidx in topology.graph.edge_indices() {
            let (a, b) = topology.graph.edge_endpoints(eidx).unwrap();
            let chan = &topology.graph[eidx];
            g.add_edge(a, b, TracedChannel::new_from(&chan));
        }
        Topology {
            graph: g,
            id2idx: topology.id2idx.clone(),
            links: topology.links.clone(),
        }
    }
}

impl Topology<EDistChannel> {
    pub(crate) fn new_edist(topology: &Topology<TracedChannel>) -> Self {
        // CORRECTNESS: For nodes and edges, `petgraph` guarantees that the
        // iteration order matches the order of indices.
        let mut g = DiGraph::new();
        for node in topology.graph.node_weights() {
            g.add_node(node.clone());
        }
        for eidx in topology.graph.edge_indices() {
            let (a, b) = topology.graph.edge_endpoints(eidx).unwrap();
            let chan = &topology.graph[eidx];
            g.add_edge(a, b, EDistChannel::new_from(&chan));
        }
        Topology {
            graph: g,
            id2idx: topology.id2idx.clone(),
            links: topology.links.clone(),
        }
    }
}

#[derive(Debug, thiserror::Error)]
pub enum TopologyError {
    #[error("Duplicate node ID {0}")]
    DuplicateNodeId(NodeId),

    #[error("Node {0} is connected to itself")]
    NodeAdjacentSelf(NodeId),

    #[error("Node {0} is not declared")]
    UndeclaredNode(NodeId),

    #[error("Duplicate links between {n1} and {n2}")]
    DuplicateLink { n1: NodeId, n2: NodeId },

    #[error("Host {id} has too many links (expected 1, got {n})")]
    TooManyHostLinks { id: NodeId, n: usize },

    #[error("Node {0} is not connected to any other node")]
    IsolatedNode(NodeId),
}

#[cfg(test)]
mod tests {
    use anyhow::Context;

    use super::*;
    use crate::testing;
    use crate::units::{Gbps, Nanosecs};

    #[test]
    fn empty_topology_succeeds() {
        assert!(
            Topology::<Channel>::new(&[], &[]).is_ok(),
            "failed to create empty topology"
        );
    }

    #[test]
    fn three_node_topology_works() -> anyhow::Result<()> {
        let (nodes, links) = testing::three_node_config();
        let topo = Topology::new(&nodes, &links).context("failed to create topology")?;
        insta::assert_yaml_snapshot!(topo.graph);
        Ok(())
    }

    #[test]
    fn eight_node_topology_works() -> anyhow::Result<()> {
        let (nodes, links) = testing::eight_node_config();
        let topo = Topology::<Channel>::new(&nodes, &links).context("failed to create topology")?;
        insta::assert_yaml_snapshot!(topo.graph);
        Ok(())
    }

    #[test]
    fn duplicate_node_fails() {
        let n1 = Node::new_host(NodeId::new(0));
        let n2 = Node::new_host(NodeId::new(0)); // error
        let n3 = Node::new_switch(NodeId::new(2));
        let l1 = Link::new(n1.id, n3.id, Gbps::default(), Nanosecs::default());
        let l2 = Link::new(n2.id, n3.id, Gbps::default(), Nanosecs::default());
        let res = Topology::new(&[n1, n2, n3], &[l1, l2]);
        assert!(matches!(res, Err(TopologyError::DuplicateNodeId(..))));
    }

    #[test]
    fn node_adjacent_self_fails() {
        let n1 = Node::new_host(NodeId::new(0));
        let n2 = Node::new_host(NodeId::new(1));
        let n3 = Node::new_switch(NodeId::new(2));
        let l1 = Link::new(n1.id, n3.id, Gbps::default(), Nanosecs::default());
        let l2 = Link::new(n2.id, n3.id, Gbps::default(), Nanosecs::default());
        let l3 = Link::new(n3.id, n3.id, Gbps::default(), Nanosecs::default()); // error
        let res = Topology::new(&[n1, n2, n3], &[l1, l2, l3]);
        assert!(matches!(res, Err(TopologyError::NodeAdjacentSelf(..))));
    }

    #[test]
    fn undeclared_node_fails() {
        let n1 = Node::new_host(NodeId::new(0));
        let n2 = Node::new_host(NodeId::new(1));
        let n3 = Node::new_switch(NodeId::new(2));
        let l1 = Link::new(n1.id, n3.id, Gbps::default(), Nanosecs::default());
        let l2 = Link::new(n2.id, n3.id, Gbps::default(), Nanosecs::default());
        let l3 = Link::new(NodeId::new(3), n3.id, Gbps::default(), Nanosecs::default());
        let res = Topology::new(&[n1, n2, n3], &[l1, l2, l3]);
        assert!(matches!(res, Err(TopologyError::UndeclaredNode(..))));
    }

    #[test]
    fn duplicate_links_fails() {
        let n1 = Node::new_host(NodeId::new(0));
        let n2 = Node::new_host(NodeId::new(1));
        let n3 = Node::new_switch(NodeId::new(2));
        let l1 = Link::new(n1.id, n3.id, Gbps::default(), Nanosecs::default());
        let l2 = Link::new(n2.id, n3.id, Gbps::default(), Nanosecs::default());
        let l3 = Link::new(n2.id, n3.id, Gbps::default(), Nanosecs::default()); // error
        let res = Topology::<Channel>::new(&[n1, n2, n3], &[l1, l2, l3]);
        assert!(matches!(res, Err(TopologyError::DuplicateLink { .. })));
    }

    #[test]
    fn too_many_host_links_fails() {
        let n1 = Node::new_host(NodeId::new(0));
        let n2 = Node::new_host(NodeId::new(1));
        let n3 = Node::new_switch(NodeId::new(2));
        let n4 = Node::new_switch(NodeId::new(3));
        let l1 = Link::new(n1.id, n3.id, Gbps::default(), Nanosecs::default());
        let l2 = Link::new(n2.id, n3.id, Gbps::default(), Nanosecs::default());
        let l3 = Link::new(n1.id, n4.id, Gbps::default(), Nanosecs::default()); // error
        let res = Topology::new(&[n1, n2, n3, n4], &[l1, l2, l3]);
        assert!(matches!(
            res,
            Err(TopologyError::TooManyHostLinks { n: 2, .. })
        ));
    }

    #[test]
    fn isolated_node_fails() {
        let n1 = Node::new_host(NodeId::new(0));
        let n2 = Node::new_host(NodeId::new(1));
        let n3 = Node::new_switch(NodeId::new(2));
        let n4 = Node::new_host(NodeId::new(3)); // error
        let l1 = Link::new(n1.id, n3.id, Gbps::default(), Nanosecs::default());
        let l2 = Link::new(n2.id, n3.id, Gbps::default(), Nanosecs::default());
        let res = Topology::new(&[n1, n2, n3, n4], &[l1, l2]);
        assert!(matches!(res, Err(TopologyError::IsolatedNode(..))));
    }

    #[test]
    fn new_topo_old_topo_equiv() -> anyhow::Result<()> {
        let (nodes, links) = testing::eight_node_config();
        let topo1 =
            Topology::<Channel>::new(&nodes, &links).context("failed to create topology")?;
        let topo2 = Topology::new_traced(&topo1);
        // Iteration order matches the order of indices
        for (n1, n2) in topo1.graph.node_weights().zip(topo2.graph.node_weights()) {
            assert_eq!(n1, n2);
        }
        for (e1, e2) in topo1.graph.edge_weights().zip(topo2.graph.edge_weights()) {
            let e2 = &Channel::new(e2.src, e2.dst, Gbps::new(100), Nanosecs::new(1000));
            assert_eq!(e1, e2);
        }
        Ok(())
    }
}
