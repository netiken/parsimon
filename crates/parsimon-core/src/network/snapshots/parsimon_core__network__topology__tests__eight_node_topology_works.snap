---
source: crates/parsimon-core/src/network/topology.rs
expression: topo.graph
---
nodes:
  - id: 0
    kind: Host
  - id: 1
    kind: Host
  - id: 2
    kind: Host
  - id: 3
    kind: Host
  - id: 4
    kind: Switch
  - id: 5
    kind: Switch
  - id: 6
    kind: Switch
  - id: 7
    kind: Switch
node_holes: []
edge_property: directed
edges:
  - - 0
    - 4
    - src: 0
      dst: 4
      bandwidth: 100000000000
      delay: 1000
  - - 4
    - 0
    - src: 4
      dst: 0
      bandwidth: 100000000000
      delay: 1000
  - - 1
    - 4
    - src: 1
      dst: 4
      bandwidth: 100000000000
      delay: 1000
  - - 4
    - 1
    - src: 4
      dst: 1
      bandwidth: 100000000000
      delay: 1000
  - - 2
    - 5
    - src: 2
      dst: 5
      bandwidth: 100000000000
      delay: 1000
  - - 5
    - 2
    - src: 5
      dst: 2
      bandwidth: 100000000000
      delay: 1000
  - - 3
    - 5
    - src: 3
      dst: 5
      bandwidth: 100000000000
      delay: 1000
  - - 5
    - 3
    - src: 5
      dst: 3
      bandwidth: 100000000000
      delay: 1000
  - - 4
    - 6
    - src: 4
      dst: 6
      bandwidth: 100000000000
      delay: 1000
  - - 6
    - 4
    - src: 6
      dst: 4
      bandwidth: 100000000000
      delay: 1000
  - - 4
    - 7
    - src: 4
      dst: 7
      bandwidth: 100000000000
      delay: 1000
  - - 7
    - 4
    - src: 7
      dst: 4
      bandwidth: 100000000000
      delay: 1000
  - - 5
    - 6
    - src: 5
      dst: 6
      bandwidth: 100000000000
      delay: 1000
  - - 6
    - 5
    - src: 6
      dst: 5
      bandwidth: 100000000000
      delay: 1000
  - - 5
    - 7
    - src: 5
      dst: 7
      bandwidth: 100000000000
      delay: 1000
  - - 7
    - 5
    - src: 7
      dst: 5
      bandwidth: 100000000000
      delay: 1000

